File: cases/taint/scala/recursive-exception-persistence-negative/RecursiveExceptionPersistenceNegative.scala
package dataflowbench

object RecursiveExceptionPersistenceNegative {
  class Box(var value: Int = 0)

  private class RecursiveSignal extends Exception

  def dfb_source(): Int = { // DFB-SOURCE: recursive-exception-persistence-input
    7
  }

  def dfb_sink(value: Int): Unit = {} // DFB-SINK: recursive-exception-persistence-sink

  def walk(box: Box, value: Int, depth: Int): Unit = {
    if (depth == 0) {
      box.value = value // DFB-WITNESS: recursive-exception-persistence-base
      box.value = 0 // DFB-KILL: recursive-exception-persistence-base-clean
      throw new RecursiveSignal // DFB-WITNESS: recursive-exception-persistence-throw
    } else {
      walk(box, value, depth - 1) // DFB-WITNESS: recursive-exception-persistence-transfer
    }
  }

  def run(): Unit = {
    val box = new Box()
    try {
      walk(box, dfb_source(), 3)
    } catch {
      case caught: RecursiveSignal => // DFB-WITNESS: recursive-exception-persistence-catch
        dfb_sink(box.value + 1) // DFB-WITNESS: recursive-exception-persistence-compose
    }
  }
}

File: cases/taint/go/recursive-exception-persistence-negative/recursive_exception_persistence_negative.go
package dataflowbench

type FlowBox struct {
	value int
}

// recursiveFlowSignal is private to this fixture and is the only recoverable signal.
type recursiveFlowSignal struct{}

func dfb_source() int { // DFB-SOURCE: recursive-exception-persistence-input
	return 7
}

func dfb_sink(value int) {} // DFB-SINK: recursive-exception-persistence-sink

func walk(box *FlowBox, value int, depth int) {
	if depth == 0 {
		box.value = value            // DFB-WITNESS: recursive-exception-persistence-base
		box.value = 0                // DFB-KILL: recursive-exception-persistence-base-clean
		panic(recursiveFlowSignal{}) // DFB-WITNESS: recursive-exception-persistence-throw
	}
	walk(box, value, depth-1) // DFB-WITNESS: recursive-exception-persistence-transfer
}

func run() {
	box := &FlowBox{value: 0}
	defer func() {
		recovered := recover()
		if _, ok := recovered.(recursiveFlowSignal); !ok {
			panic(recovered)
		}
		dfb_sink(box.value + 1) // DFB-WITNESS: recursive-exception-persistence-compose
	}()
	walk(box, dfb_source(), 3)
}
